File: project/cloud/models/anti/AfBase.php
<?php

namespace app\models\anti;

use Yii;

/**
 * This is the model class for table "af_base".
 *
 * @property string $id
 * @property integer $request_id
 * @property integer $aid
 * @property integer $user_id
 * @property integer $loan_id
 * @property integer $jxlstat_id
 * @property integer $match_status
 * @property string $modify_time
 * @property string $create_time
 */
class AfBase extends AntiBaseModel
{
    /**
     * @inheritdoc
     */
    public static function tableName()
    {
        return 'af_base';
    }

    /**
     * @inheritdoc
     */
    public function rules()
    {
        return [
            [['request_id', 'aid', 'user_id', 'loan_id', 'jxlstat_id', 'create_time'], 'required'],
            [['request_id', 'aid', 'user_id', 'loan_id', 'jxlstat_id', 'match_status'], 'integer'],
            [['modify_time', 'create_time'], 'safe'],
        ];
    }

    /**
     * @inheritdoc
     */
    public function attributeLabels()
    {
        return [
            'id' => '主键，自增',
            'request_id' => '请求处理id',
            'aid' => '业务ID',
            'user_id' => '用户ID',
            'loan_id' => '贷款ID',
            'jxlstat_id' => '聚信立stat表的id',
            'match_status' => '二级匹配状态:0:初始; 1:锁定; 2:成功',
            'modify_time' => '修改时间',
            'create_time' => '创建时间'
        ];
    }

    /**
     * 获取需要匹配间接关系的数据
     */
    public function getJaccardData($data, $field = '*')
    {
        if (empty($data)) {
            return false;
        }
        $where = [
            'AND',
            ['match_status' => $data['match_status']],
            ['>', 'create_time', $data['create_time']],
        ];
//        $query = static::find()->where($where)->select($field)->asArray()->limit(500)->orderBy('create_time desc');
//        echo $query->createCommand()->getRawSql(); exit;
        return static::find()->where($where)->select($field)->asArray()->limit(500)->orderBy('create_time desc')->all();
    }

    /**
     * 锁定为间接关系处理中状态
     */
    public function lockJcards($ids) {
        if (empty($ids)) {
            return false;
        }
        $sets = [
            'match_status' => 5,
            'modify_time' => date('Y-m-d H:i:s'),
        ];
        $where = [
            'id' => $ids,
            'match_status' => 4,
        ];
        $result = static::updateAll($sets, $where);
        return $result;
    }

    /**
     * 更新为结束状态
     */
    public function finishJcard($id) {
        if (empty($id)) {
            return false;
        }
        $sets = [
            'match_status' => 6,
            'modify_time' => date('Y-m-d H:i:s'),
        ];
        $where = [
            'id' => $id
        ];
        $result = static::updateAll($sets, $where);
        return $result;
    }
}

File: project/graphdata/model/antifraud/af_base.py
# -*- coding: utf-8 -*-
from lib.application import db
from .base import Base
from datetime import datetime, timedelta
from sqlalchemy import and_

class AfBase(db.Model, Base):
    # 指定数据库
    __bind_key__ = 'xhh_antifraud'
    __tablename__ = 'af_base'

    id = db.Column(db.BigInteger, primary_key=True)
    request_id = db.Column(db.BigInteger, nullable=False, index=True)
    aid = db.Column(db.Integer,nullable=False)
    user_id = db.Column(db.BigInteger, nullable=False, index=True)
    loan_id = db.Column(db.BigInteger, nullable=False, index=True)
    jxlstat_id = db.Column(db.BigInteger, nullable=False, index=True)
    match_status = db.Column(db.Integer, nullable=False)
    create_time = db.Column(db.DateTime, nullable=False)
    modify_time = db.Column(db.DateTime, nullable=False)

    def getById(self, id):
        db_base = self.query.get(int(id))
        return db_base

    def save(self):
        db.session.add(self)
        try:
            db.session.flush()
            db.session.commit()
        except Exception:
            db.session.rollback()
            raise

    def getMulMatchData(self, start_time):
        '''
        获取需要匹配二级关系的数据
        '''
        where = and_(
            AfBase.create_time >= start_time,
            AfBase.match_status == 2
        )

        res = db.session.query(AfBase)   \
            .filter(where) \
            .order_by(AfBase.create_time) \
            .limit(500) \
            .all()

        return res

    def getJaccardData(self, start_time):
        '''
        获取需要匹配间接关系(jaccard)的数据
        '''
        where = and_(
            AfBase.create_time >= start_time,
            AfBase.match_status == 4
        )

        res = db.session.query(AfBase) \
            .filter(where) \
            .order_by(AfBase.create_time) \
            .limit(500) \
            .all()

        return res
    def getRelationData(self, start_time):
        '''
        获取已存在关系(relation)的数据
        '''
        where = and_(
            AfBase.create_time >= start_time,
            AfBase.match_status == 6
        )

        res = db.session.query(AfBase) \
            .filter(where) \
            .order_by(AfBase.create_time) \
            .limit(500) \
            .all()

        return res
    def getJaccardDataByids(self, ids):
        '''
        获取需要匹配间接关系(jaccard)的数据
        '''
        where = and_(
            AfBase.match_status == 4
        )
        res = db.session.query(AfBase) \
            .filter(AfBase.match_status.in_([4,5,6])) \
            .order_by(AfBase.create_time) \
            .group_by(AfBase.user_id) \
            .filter(AfBase.user_id.in_(ids)) \
            .limit(500) \
            .all()
        return res

    def getAfBaseData(self, start_time):
        where = and_(
            AfBase.create_time >= start_time, 
            AfBase.match_status == 0, #初始
            # AfBase.aid == aid
        )

        res = db.session.query(AfBase)   \
            .filter(where) \
            .order_by(AfBase.create_time) \
            .limit(1000) \
            .all()

        return res

    def lockMatchStatus(self, data, status):
        if len(data) == 0:
            return False

        now = datetime.now()
        for o in data:
            o.modify_time = now
            o.match_status = status  # 锁定 | 完成

        db.session.commit()

        return True

    def finishMatched(self, match_status):
        '''
        结束二级关系匹配流程
        '''
        self.match_status = match_status
        self.modify_time = datetime.now()
        db.session.commit()
